server/controllers/userController.js

```diff
@@ -0,0 +1,88 @@
+const bcrypt = require("bcrypt");
+const jwt = require("jsonwebtoken");
+const passport = require("passport");
+const { body, validationResult } = require("express-validator");
+const User = require("../models/User");
+
+exports.create = [
+  body("username")
+    .notEmpty()
+    .withMessage("Username is required")
+    .custom(async (value) => {
+      const user = await User.findOne({ username: value });
+      return user
+        ? Promise.reject("Username already in use.")
+        : Promise.resolve(true);
+    })
+    .trim(),
+  body("email")
+    .notEmpty()
+    .withMessage("Email is required")
+    .isEmail()
+    .withMessage("Must be valid email address")
+    .custom(async (value) => {
+      const user = await User.findOne({ email: value });
+      return user
+        ? Promise.reject("Email already in use.")
+        : Promise.resolve(true);
+    })
+    .trim()
+    .normalizeEmail(),
+  body("password")
+    .isLength({ min: 8, max: 60 })
+    .withMessage("Passwords must be at least 8 characters long.")
+    .trim(),
+  async (req, res, next) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({
+        errors: errors.array(),
+      });
+    }
+    try {
+      const { username, password, email } = req.body;
+      const hashedPassword = await bcrypt.hash(password, 10);
+      const user = new User({
+        username: username,
+        email,
+        password: hashedPassword,
+      });
```

Comment:
Using a [Mongoose setter](https://mongoosejs.com/docs/tutorials/getters-setters.html#setters) to hash the password instead of doing it here will help clean up the controller a bit and separate the concern of password hashing from Express (in case you, for example, wanted to use those models with another service).

Reply:
Hmm couldn't seem to get the setter to work so I opted for a pre save function instead. The password hashing is at least done in the user model now.